File: src/main/java/linkedList/LinkedListCycle.java
package linkedList;

/**
 * @author Max DeMaio
 * Solved on 06/03/2022
 *
 * Linked List Cycle
 * https://leetcode.com/explore/interview/card/top-interview-questions-easy/93/linked-list/773/
 * Difficulty: Easy
 *
 * Approach: Tortoise and hare
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 *
 */

public class LinkedListCycle {
    public boolean hasCycle(ListNode head) {
        if (head == null) {
            return false;
        }


        ListNode slow = head;
        ListNode fast = head.next;

        while (slow != fast) {
            if (fast == null || fast.next == null) {
                return false;
            }

            slow = slow.next;
            fast = fast.next.next;
        }

        return true;
    }
}

File: python_problems/linkedlist/linkedlistcycle.py
from typing import Optional
from linkedlist.util import ListNode

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False
        
        tortoise = head
        hare = head.next

        while tortoise != hare:
            if hare == None or hare.next == None:
                return False
            tortoise = tortoise.next
            hare = hare.next.next
        return True
